lib/services/crawler/crawler.go

```diff
@@ -0,0 +1,153 @@
+	}
+
+	c.Keyword.Status = models.Processing
+	err := c.Keyword.Save(c.DB)
+	if err != nil {
+		c.Keyword.Status = models.Failed
+		_ = c.Keyword.Save(c.DB)
+		return err
+	}
+
+	parsingResult := ParsingResult{}
+	if c.Collector == nil {
+		c.Collector = colly.NewCollector()
+	}
+
+	extensions.RandomUserAgent(c.Collector)
+
+	c.Collector.OnRequest(func(r *colly.Request) {
+		log.Infof("Visiting: %v user-agent: %v", r.URL, r.Headers.Get("User-Agent"))
+	})
+
+	c.Collector.OnResponse(func(r *colly.Response) {
+		parsingResult.HTMLCode = string(r.Body[:])
+	})
+
+	c.Collector.OnError(func(r *colly.Response, err error) {
+		log.Errorf("Request URL: %v failed with error: %v", r.Request.URL, err)
+		c.Keyword.Status = models.Failed
+		_ = c.Keyword.Save(c.DB)
+	})
+
+	for _, pattern := range selectors["nonAds"] {
+		c.Collector.OnHTML(pattern, func(e *colly.HTMLElement) {
+			link := formatLink(e.Attr("href"))
+			parsingResult.NonAdwordLinks = append(parsingResult.NonAdwordLinks, link)
+		})
+	}
+
+	for _, pattern := range selectors["ads"] {
+		c.Collector.OnHTML(pattern, func(e *colly.HTMLElement) {
+			link := formatLink(e.Attr("href"))
+			parsingResult.AdwordLinks = append(parsingResult.AdwordLinks, link)
+		})
+	}
+
+	c.Collector.OnHTML(selectors["mobileLinks"][0], func(e *colly.HTMLElement) {
+		link := formatLink(e.Attr("href"))
+		if len(e.DOM.Find(selectors["mobileAds"][0]).Nodes) > 0 {
+			parsingResult.AdwordLinks = append(parsingResult.AdwordLinks, link)
+		} else {
+			parsingResult.NonAdwordLinks = append(parsingResult.NonAdwordLinks, link)
+		}
+	})
+
+	c.Collector.OnScraped(func(r *colly.Response) {
+		c.parsingResult = &parsingResult
+		err := c.save()
+		if err != nil {
+			log.Error("Error when saving to DB:", err)
+
+			c.Keyword.Status = models.Failed
+			err = c.Keyword.Save(c.DB)
+			if err != nil {
+				log.Error("Error, cannot update status to Failed:", err)
+			}
+		}
+	})
+
+	requestUrl := fmt.Sprintf(urlPattern, url.QueryEscape(c.Keyword.Text))
+	return c.Collector.Visit(requestUrl)
+}
+
+func (c *Crawler) save() error {
+	nonAdwordLinks, err := json.Marshal(c.parsingResult.NonAdwordLinks)
+	if err != nil {
+		return err
+	}
+
+	adwordLinks, err := json.Marshal(c.parsingResult.AdwordLinks)
+	if err != nil {
+		return err
+	}
+
+	nonAdwordLinksCount := len(c.parsingResult.NonAdwordLinks)
+	adwordLinksCount := len(c.parsingResult.AdwordLinks)
+	shopAdwordLinksCount := len(c.parsingResult.ShopAdwordLinks)
+	totalCount := nonAdwordLinksCount + adwordLinksCount + shopAdwordLinksCount
+
+	keyword := c.Keyword
+	keyword.NonAdwordLinksCount = nonAdwordLinksCount
+	keyword.NonAdwordLinks = sql.NullString{String: string(nonAdwordLinks)}
+	keyword.AdwordLinksCount = adwordLinksCount
+	keyword.AdwordLinks = sql.NullString{String: string(adwordLinks)}
+	keyword.LinksCount = totalCount
+	keyword.HtmlCode = c.parsingResult.HTMLCode
+
+	keyword.Status = models.Completed
+
+	return keyword.Save(c.DB)
```

Comment:
This logic should be separated into a new struct

Reply:
Same as above [comment](https://github.com/markgravity/golang-ic/pull/43#discussion_r1309747194) 🙏